include/swift/AST/DiagnosticEngine.h

```diff
@@ -870,6 +888,11 @@ namespace swift {
       if (IsOpen) {
         commit();
       }
+
+      if (Depth == 0) {
+        Engine.TransactionStrings.clear();
+        Engine.TransactionAllocator.Reset();
```

Comment:
I had originally thought to put the BumpPtrAllocator in the transaction RAII object, but this works too.

Reply:
Yes, that's what I did originally but then I realized that diagnostics are not emitted until the last transaction in the chain is complete, so decided to put it into diagnostic engine itself.

Reply:
Oops, right, of course. That's the whole point of nested transactions.